libursa/Database.cpp

```diff
@@ -69,24 +69,46 @@ void Database::create(const std::string &fname) {
     empty.save();
 }
 
+bool Database::can_acquire(const DatasetLock &newlock) const {
+    for (const auto &[k, task] : tasks) {
+        if (task->locks_dataset(newlock.target())) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool Database::can_acquire(const IteratorLock &newlock) const {
+    for (const auto &[k, task] : tasks) {
+        if (task->locks_iterator(newlock.target())) {
+            return false;
+        }
+    }
+    return true;
+}
+
 uint64_t Database::allocate_task_id() { return ++last_task_id; }
 
-Task *Database::allocate_task(const std::string &request,
-                              const std::string &conn_id) {
-    while (true) {
-        uint64_t task_id = allocate_task_id();
-        auto timestamp = std::chrono::steady_clock::now().time_since_epoch();
-        uint64_t epoch_ms = get_milli_timestamp();
-        if (tasks.count(task_id) == 0) {
-            return tasks
-                .emplace(task_id, std::make_unique<Task>(Task(
-                                      task_id, epoch_ms, request, conn_id)))
-                .first->second.get();
+TaskSpec *Database::allocate_task(const std::string &request,
+                                  const std::string &conn_id,
+                                  const std::vector<DatabaseLock> &locks) {
+    for (const auto &lock : locks) {
+        spdlog::info("Locking {}...", describe_lock(lock));
+        if (!std::visit([this](auto &l) { return can_acquire(l); }, lock)) {
+            spdlog::warn("Can't lock {}!", describe_lock(lock));
+            throw std::runtime_error("Can't acquire lock");
         }
     }
+    uint64_t task_id = allocate_task_id();
+    auto timestamp = std::chrono::steady_clock::now().time_since_epoch();
+    uint64_t epoch_ms = get_milli_timestamp();
+    return tasks
+        .emplace(task_id, std::make_unique<TaskSpec>(task_id, conn_id, request,
+                                                     epoch_ms, locks))
+        .first->second.get();
```

Comment:
Can we make it more readable and split it into 2 parts, namely emplacing and extracting pointer